---
title: "Estimate MCN using CASMCN"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Estimate-MCN-using-CASMCN}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

This tutorial shows how to estimate relative mitochondrial copy number (MCN) from CEL files generated by CAS Array using CASMCN package. For more information about CAS Array, please refer to [our article]. For the methods about MCN estimation, please refer to [the github page](https://github.com/Zijian-Tian/CASMCN).

# System Requirements

In addition to the installation of the CASMCN package, [APT](https://www.thermofisher.com/cn/zh/home/life-science/microarray-analysis/microarray-analysis-partners-programs/affymetrix-developers-network/affymetrix-power-tools.html) and [PennCNV](https://penncnv.openbioinformatics.org/en/latest/) are also needed for MCN estimation. The pipeline was tested on APT 2.11.4 and PennCNV 1.0.5, but should also work fine on other versions.

# Input Files

## CEL Files

Genotyping with CAS Array produces a CEL file for each sample. We use a character vector to locate all CEL files included in MCN estimation:

```{r, eval = FALSE}
cel_list <- c("/path/to/sample1.CEL",
              "/path/to/sample2.CEL",
              "/path/to/sample3.CEL",
              ...)
```

## Library Files

APT needs a set of library file to process CEL files. Library files of CAS Array could be obtained from manufacturer. The extracted library files should contain multiple .xml parameter files, a .annot.csv annotation file and several other files essential for APT processing. 

Most library files needed by PennCNV will be generaged from the annotation file of the CAS Array, but the gc model file need to be specified additionally. Since the genome build of the library files of CAS Array is hg38, the gc model file needs to be the hg38 version. You could find the file in gc_file/hg38.gc5Base.txt.gz in your local PennCNV installation, or download the file from [the github repository of PennCNV](https://github.com/WGLab/PennCNV/blob/master/gc_file/hg38.gc5Base.txt.gz).

```{r, eval = FALSE}
apt_lib_dir <- "/path/to/APT_library_directory"
gc5_file <- "/path/to/hg38.gc5Base.txt.gz"
```

# Basic Usage

The minimal requirements to run MCN estimation is the input files above. To start the analysis, use the function `estimate_mcn_from_cel`:

```{r, eval = FALSE}
library(CASMCN)
estimate_mcn_from_cel(cel_list,
                      output_dir = "/path/to/output_directory",
                      apt_lib_dir,
                      gc5_file)
```

This function will generate relative MCN estimates for the samples and save the result to "CASMCN_result.txt" in your `output_dir`. A log file "estimate_MCN_from_CEL.log" will also be generated for tracking the progress since the analysis may take up to a few hundred of hours when applied to larger sample size. 

In addition to the basic options, **we recommend to provide a phenotype associated to MCN and choose an optimal number of PCs used (see below)** for the PC oriantation step, especially when the sample size is small (e.g. <1000).

# Output Format

The final result of MCN estimation will be written to "CASMCN_result.txt" in the output directory. It is a tab-delimited text file with two columns "Sample_Name" and "MCN". The sample name is the file name of each CEL file. MCN is the estimated relative mitochondrial copy number. **Note that the estimated MCN is not the absolute copy number of mitochondria. The relative MCN estimates could not be compared between different populations.**

# Optional Arguments

## Associated Phenotype

An external phenotype associated with MCN is needed to oriantate the PC derived from mitochondrial intensity signal. The correlated phenotype should be provided as a data frame with two columns "Sample_Name" and "Phenotype":

```{r, include = FALSE}
correlated_pheno <- data.frame(Sample_Name = paste0("sample", seq(6), ".CEL"),
                               Phenotype = rnorm(6))
```

```{r}
head(correlated_pheno)
```

The phenotype should have known correlation with MCN. Commonly used traits includes white blood cell counts (negative correlation), red blood cel counts (negative correlation), platelet count (positive correlation), age (negative correlation) and sex (female > male). If you have available MCN estimates for part of the samples, that also could be used in this step (as a phenotype positively associted with MCN). The direction of the association need to be clarified by `correlation_direction`:

```{r, eval = FALSE}
correlation_direction <- "+"  # phenotype positively correlated with MCN
# OR
correlation_direction <- "-"  # phenotype negatively correlated with MCN
```

If no external phenotype is provided (`correlated_pheno = NULL`), PC oriantation will be done with a polygenic score (PGS) of MCN derived from a [GWAS of MCN](https://doi.org/10.1007/s00439-021-02394-w) using [LDpred2](https://doi.org/10.1093/bioinformatics/btaa1029). Since the SNP heritability of MCN is not high enough, PGS may not powerful enough for PC oriantation in small population (e.g. <1000). **We recommend to provide external phenotype whenever it is possible.**

## Autosomal Principal Components Included

We use principal components (PCs) of intensity signal from autosomal markers to capture confounding factors including batch effects and DNA concentrations. If the number of PCs is not enough to capture the confounders, the result is unreliable. Empirically, 1% of the sample size is enough for MCN estimation in large populations. For MCN estimation in smaller populations, number of PCs may need to be set manually.

To control the number of PCs included, you can set `pc_used` in two ways: By proportion of the sample size passed the previous quality control (a fraction between 0 and 1); Or by the exact number of PCs included (a positive integer).

```{r, eval = FALSE}
pc_used <- 0.02  # By proportion of sample size
# OR
pc_used <- 20    # By exact number of PCs
```

To determine how many PCs should be used, **we recommend to check the MCN adjusted with different number of PCs by `plot_pc_adjustment` function**. This function takes the `output_dir` of `estimate_mcn_from_cel` as the input and use its intermediate results to plot correlation between estimated MCN and external phenotype (or MCN PGS if no phenotype is provided) against number of PCs used for adjustment. A typical plot is as follows:

```{r, eval = FALSE}
plot_pc_adjustment(output_dir)
```

```{r, echo = FALSE}
# plot the pre-computed result
plot(CASMCN:::pc_selection_plot, xlab = "Number of PC(s) included", ylab = "Correlation between MCN estimates and MCN_PGS")
```

The association between estimated MCN and external phenotype become stable after adjusted for ~100 PCs, giving a suggestion for the PC number selection. 

The `pc_used` parameter of the `estimate_mcn_from_cel` function can be adjusted safely while using existing temporary files from previous runs: With full intermediate results, you can just adjust the `pc_used` and re-run `estimate_mcn_from_cel` to generate a new MCN estimate. This should be fast if the new `pc_used` is smaller than the previous one since all necessery information could be loaded from the intermediate results. If the specified number of PCs used is larger than the cached PCs, another PCA had to be done. Since PCA is a time-consuming process, set a larger number of PCs used in the first run is recommended. That will cache enough information for further analysis, including determine the optimal number of PCs by `plot_pc_adjustment` and estimate MCN using `estimate_mcn_from_cel` with different settings of `pc_used`.

## High Quality Markers

A set of high quality markers for MCN estimate are specified by `autosomal_markers` and `mitochondrial_markers`. By default, a set of pre-defined high quality markers is used: 

```{r, include = FALSE}
hq_autosomal_markers <- CASMCN:::hq_autosomal_markers
hq_mitochondrial_markers <- CASMCN:::hq_mitochondrial_markers
```

```{r}
str(hq_autosomal_markers)
str(hq_mitochondrial_markers)
```

To generate the pre-defined high quality markers, we applied the following filters: Exclude markers with off-target probes (more than one alignment with percentage of identical matches > 80%) for autosomal and mitochondrial markers; For autosomal markers, additional quality controls including call rate > 95%, HWE p-value > 10<sup>-6</sup>, LD r<sup>2</sup> < 0.3 and maximum spacing were done. 

## Temporary Files

This pipeline will generate a series of temporary files since multiple external softwares are included. All temporary files will be put in the "tempdir\" directory in the output directory. If `keep_tempfile` is set to `FALSE`, then all temporary files will be deleted after the final MCN estimates are generated. 

The pipeline could utilize legacy temporary files generate by previous runs. Unless `ignore_existing_tempfiles` is set to `TRUE`, the pipeline will skip the substeps as long as it detected the output of that substep. If that intermediate result is incomplete, this may cause issues. You can delete the output of the last substep and re-run the pipeline to solve the issue. If `ignore_existing_tempfiles` is set to `TRUE`, all temporary files will be deleted before the pipeline starts.

## External Software Executables

By default, the executables of APT and PennCNV will be searched from system `PATH`. You can also specify the main directory of APT and PennCNV installation explicitly by `apt_exec_dir` and `penncnv_exec_dir` arguments.